Add Network configuration objects content to the Traffic management concept

rcaballeromx opened the pull request:
The Network configuration objects concept provides a high-level overview of
the available network configuration capabilities of Istio. Cross-references,
diagrams, and examples provide a clear introduction to the objects with a clear
didactic approach and simple content structure. Links to and from existing
content help readers connect the different pieces of information into single
cognitive units. To aid that goal, I performed some consistency edits of
existing content.

Signed-off-by: rcaballeromx <grca@google.com>

istio-testing commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To fully approve this pull request, please assign additional approvers.
We suggest the following additional approver: **rcaballeromx**

If they are not already assigned, you can assign the PR to them by writing `/assign @rcaballeromx` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details open>
Needs approval from an approver in each of these files:

- **[content/OWNERS](https://github.com/istio/istio.io/blob/master/content/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["rcaballeromx"]} -->

oaktowner commented on content/docs/concepts/network/index.md:
```diff
@@ -0,0 +1,437 @@
+The following diagram shows the configured routes for the external HTTPS
+traffic coming from the external host on port 443 to the `virtual-svc` virtual
+service through the `ext-host-gwy` gateway to reach the `my-svc` service
+within the mesh:
```
I think that this following diagram shows what the previous 2 YAMLs describe. Again: probably better to start with the prose description of what we want (e.g., "traffic from outside my mesh addressed to https:my-svc should get routed to the service (workload?) called my-svc").

And it would be much better to have something like "shopping-cart" than "my-svc"

rcaballeromx replied:
That is exactly what it shows, I assume the text doesn't make that clear?
The diagrams are an effort to "show" rather than "tell" using prose. I've added a brief sentence though. 
The service can be called whatever but using "my-svc" and similar names makes it clear that these are placeholder names.

rshriram commented:
This is actually pretty cool, especially given that you are not the subject matter expert in the area. I like the intent and the layout, love the diagrams that give a quick visual overview of the networking config. Glad you could glean so much info from the networking API docs. Thank you for doing this!

All said, there are a lot of inaccuracies here, which I can understand. I strongly suggest you let @frankbu incorporate the layout and the content into the existing traffic management doc. Its important to convey the concepts to users before the mechanisms. The traffic management doc does just that (describing our design philosophies for canary rollouts, etc.) and then takes the user through the networking config.

oaktowner commented on content/docs/concepts/network/index.md:
```diff
@@ -0,0 +1,437 @@
+
+Use egress gateways to configure an exit route for traffic leaving the mesh.
+You can configure each egress gateway to use its own policies and telemetry.
+You can use egress gateways if only particular application services can or
```
Isn't there a 'must' here? That is, you _must_ use egress gateways if you want to permit traffic off the mesh?

frankbu replied:
no, sending egress traffic via a gateway is optional.

oaktowner replied:
ack.

frankbu commented on content/docs/concepts/network/index.md:
```diff
@@ -0,0 +1,437 @@
+workloads inside the mesh without a virtual service. The Istio virtual services
+can configure traffic routes to:
+
+-  Other virtual services in the mesh.
```
No they can't

rcaballeromx replied:
Understood.

frankbu commented on content/docs/concepts/network/index.md:
```diff
@@ -0,0 +1,437 @@
+    >}}
+
+As you can see, virtual services are very flexible and can accommodate a wide
+variety of topologies. With destination rules, virtual services configure the
```
Destination rules are not required

rcaballeromx replied:
Understood. What functionality would you say a destination rule can add to a virtual service? What relationships are possible between the two?

frankbu commented on content/docs/concepts/network/index.md:
```diff
@@ -0,0 +1,437 @@
+service subset. You can also configure the corresponding Istio destination rule
+to determine the set of pods or VMs belonging to these subsets.
+
+You can use the Istio virtual services in a variety of ways: to route traffic
```
What do you mean by a variety of ways, given you only follow that up with two examples?

rcaballeromx replied:
I'm trying to leave it open ended to communicate that even if what the reader is doing with the configuration objects does not match 1:1 what is described here, that does not mean that it should not be possible. I added a conditional prepositional phrase, "for example" to make clear that the two examples are just that.

frankbu commented on content/docs/concepts/network/index.md:
```diff
@@ -0,0 +1,437 @@
+    - *.my-co.org
+      route:
+      - destination:
+        host: service
```
This doesn't seem like a very realistic ... You're delegating everything all my-co.org subdomains to a single monolith service

rcaballeromx replied:
Realism was not my intent. This doc is about adding clarity to how the configuration objects behave. Starting with a simple example where everything goes to a single service is meant as an introduction. Further down the doc, more complex configurations are discussed.

frankbu commented on content/docs/concepts/network/index.md:
```diff
@@ -0,0 +1,437 @@
+gateways overcome the shortcomings of the layer 4 and 6 specifications. Istio
+only configures the functions that layer 7 load balancing proxies implement
+uniformly. To use standard Istio rules to control HTTP requests as well as TCP
+traffic entering a gateway, you must bind a virtual service to the gateway. The
```
yes, all traffic entering via an ingress gateway needs to be routed somewhere using a virtual service.

oaktowner replied:
Thanks, @frankbu .  It would be _great_ to capture that.

rcaballeromx replied:
Done

frankbu commented on content/docs/concepts/network/index.md:
```diff
@@ -0,0 +1,437 @@
+{{< /text >}}
+
+Because a service entry configuration simply adds the external resource to the
+internal service registry, you must use either a virtual service or destination
```
Depending on the situation, you don't need anything more that a service entry to call an external service

oaktowner replied:
Great to know. @rcaballeromx can you capture that?

rcaballeromx replied:
Absolutely, done!

frankbu commented on content/docs/concepts/network/index.md:
```diff
@@ -0,0 +1,437 @@
+    >}}
+
+Service entries do not require virtual services to address services in the
+mesh. Use destination rules to configure routes to address services directly.
```
Not sure what this is trying to say, but I think it's wrong.

rcaballeromx replied:
It was trying to say what you stated in the previous comment only awkwardly. I've removed it.

rcaballeromx commented:
@rshriram This content is intended to provide conceptual information about the network objects. The Traffic routing doc focuses on What you can do with traffic and how to accomplish specific routing task, which is all well and good. This doc focuses on How the network configuration objects work and why they are needed. I did link the two documents to one another but merging them into one would only make the content too long and confusing.

rcaballeromx commented:
@frankbu I fully agree that the term "gateway" is overloaded. This document introduces consistency by avoiding referring to the configuration objects using the value in the `kind` key. This document is not talking about a specific configuration file with a `kind: Gateway` value, but rather describing the conceptual behavior of the gateway configuration object. Reading your comment does leave me thinking that additional disambiguation might be needed. I did not consider that the actual workload is also called "gateway". I am trying to establish a consistent pattern for these terms in this document. Using `VirtualService` instead of virtual service is very confusing because it, incorrectly, communicates that what the text refers to only applies to the example file provided.

rcaballeromx commented:
@oaktowner @frankbu @rshriram Thank you all for your comments and review! I have replied to the general comments and will be addressing the in-line comments soon.

rcaballeromx commented:
@oaktowner Regarding your raised issues about the diagrams:

1. I depicted the Envoy proxies because none of the configuration happens on the actual workloads. The configuration is always sent to the proxies. I could remove the proxies from the diagram but that could create a lot of confusion with folks thinking the configuration affects services.

1. They are neither E:R nor network diagrams. Lines 20-24 explain clearly what the diagrams are depicting.

rcaballeromx commented:
@frankbu I did follow an Istio-wide policy: https://istio.io/about/contribute/style-guide/
The case around virtual service vs `VirtualService` is the similar to Mixer vs mixer or "service mesh" vs "Service Mesh". 
The markup and spelling of things like `VirtualService` imply a connection to the actual value in the configuration file. If I was writing content that needed that direct connection, I would use it but the content here is not referring to that specific value.

rcaballeromx commented on content/docs/concepts/network/index.md:
```diff
@@ -0,0 +1,437 @@
+
+## Gateways
+
+The Istio gateways are layer 7 load balancers. With an Istio gateway, you can
```
Is it correct to say that every configuration object with `kind: Gateway` requires 1 or more workloads to work? How can I identify the workload in the configuration object? You mention that they only configure layer 4-6 but I read [here](https://istio.io/blog/2018/v1alpha3-routing/#gateway) that they implement L7. Can you help me by clarifying further?

frankbu replied:
Right, so this is the problem. A "gateway" is a workload, an L7 router implementation (typically an Envoy running in a pod). A `Gateway`, which is what you are talking about here, is the kind of the configuration resource that configures the L4-6 properties of the gateway that it is bound to. It's bound to a particular gateway (workload) with the `selector` field in the `Gateway`. A `VirtualService` is then used to configure the remaining L7 properties of the gateway. This is why I think its best to explicitly refer to the configuration resources by the value of there kind field. That seems pretty clear to me.

frankbu replied:
Btw, the style guide doesn't really address the issue of the configuration resource kinds although it does have examples of referring to them as I suggest: https://istio.io/about/contribute/style-guide/#use-code-style-for-inline-code-and-commands

Notice the word `Deployment`.

rcaballeromx replied:
@frankbu Thank you for the clarification. I've modified the text using your outstanding explanation. To avoid the terminology issue, I'm using the term "gateway configuration object" to distinguish the actual "gateway workload" 
IMO using the value of `kind:` doesn't really solve the problem of us referring to two very different things with the same term, it just transforms it into a markup use consistency issue. 
The best practice in technical writing is clear here: https://docs.microsoft.com/en-us/style-guide/word-choice/use-technical-terms-carefully
Having some specific research on this might be helpful but until then, I'll use the workaround described above.

geeknoid commented on content/docs/concepts/network/index.md:
```diff
@@ -0,0 +1,437 @@
+---
+title: The Istio network configuration objects and your application
```
This is a bit long for our normal document titles, can this be made shorter?

rcaballeromx replied:
done

geeknoid commented on content/docs/concepts/network/index.md:
```diff
@@ -0,0 +1,437 @@
+Lastly, the document discusses the benefits of using the Istio network API for
+your cloud applications.
+
+# The Istio network configuration objects
```
For our docs, the highest-level header is H2. So this should be ##

rcaballeromx replied:
Done

rcaballeromx commented on content/docs/concepts/network/index.md:
```diff
@@ -0,0 +1,437 @@
+Cloud applications, unlike monolithic applications, are comprised of a
+collection of connected services. Throughout this document a service should be
+understood as an addressable destination workload inside the mesh. These
+services can connect to other services or to external resources. The Istio
```
Shifted focus to the readers application and rewrote the two sentences to provide a better intro to the document.

rcaballeromx commented on content/docs/concepts/network/index.md:
```diff
@@ -0,0 +1,437 @@
+
+The Istio network configuration objects provide you with four objects to
+configure the traffic routes between your services and your external resources:
+virtual services, destination rules, service entries, and gateways. You can add
```
Done

rcaballeromx commented on content/docs/concepts/network/index.md:
```diff
@@ -0,0 +1,437 @@
+specific mesh, you can combine the configuration objects in a way that best
+fulfills your needs.
+
+You can define the Istio network configuration objects as Kubernetes Custom
```
I used can because, theoretically, readers could be using other constructs if they do not use Kubernetes. That being said and since the entire doc focuses heavily on CRDs, I agree with you.

rcaballeromx commented on content/docs/concepts/network/index.md:
```diff
@@ -0,0 +1,437 @@
+describe the purposes and specific configurations of each configuration object
+and provide examples.
+
+## Virtual Services
```
I followed a third approach, a didactic one. I began with the simplest and most familiar, virtual service, and went to the more complex and unfamiliar, service entries. There is a dependency between gateways and virtual services though. Because of it, I went with the virtual services first.

rcaballeromx commented on content/docs/concepts/network/index.md:
```diff
@@ -0,0 +1,437 @@
+        subset: v1
+{{< /text >}}
+
+Virtual services can specify the destination's DNS host name using the `hosts:`
```
I took another stab based on your suggestions. With the exception of your final suggestion. Overwhelming readers by mentioning all the other places where "service" is defined would defeat the main purpose of the doc, which is to teach how the configuration objects work. The configuration files in this doc are not meant to be tried out nor is anywhere implied they would "work" if the reader just copy pasted them somewhere. That type of content and examples is meant for procedures, not concepts.

rcaballeromx commented on content/docs/concepts/network/index.md:
```diff
@@ -0,0 +1,437 @@
+wildcard for the virtual service configurations to apply to multiple hosts. In
+the `destination:` field of each route, you specify the service and subset that
+the configuration targets. The example configures a route for traffic to reach
+the `v1` subset of the `my-svc` service from the `my-vtl-svc` virtual service.
```
Moving it up.

rcaballeromx commented on content/docs/concepts/network/index.md:
```diff
@@ -0,0 +1,437 @@
+
+The following diagram shows the configured traffic routes implementing a
+virtual service to configure traffic for services in a namespace for two
+distinct application services based on the URI prefixes:
```
Yes.

rcaballeromx commented on content/docs/concepts/network/index.md:
```diff
@@ -0,0 +1,437 @@
+Once you have identified and configured the virtual services suitable for your
+mesh, you can add gateways to route traffic in or out of your mesh. The virtual
+services together with destination rules allow you to configure the behavior of
+the traffic to fulfill the needs of your mesh. Together with a service entry,
```
No, they are introduced here as the piece a virtual service needs to "configure traffic routes to external dependencies". It's an ancient didactic technique. I'm using something the reader now knows, virtual services, to begin introducing the unknown, gateways, destination rules, and service entries.
I'm adding links to the the different sections on this doc covering the mentioned objects. Hopefully it does not end up feeling too link-cluttered.

rcaballeromx commented on content/docs/concepts/network/index.md:
```diff
@@ -0,0 +1,437 @@
+  name: my-namespace
+spec:
+  hosts:
+    - *.my-namespace.svc.cluster.local
```
@frankbu Please correct me if I'm wrong but, the way this is currently configured, means:
1. Only requests intended for the `my-namespace.svc.cluster.local` namespace will follow the route in this virtual service regardless of whether they are intended for svc1 or svc2.
2. Only requests intended for the `svc1.my-namespace.svc.cluster.local` follow the route to `host: svc-1`
3. Only requests intended for the `svc2.my-namespace.svc.cluster.local` follow the route to `host: svc-2`
Just to reiterate, this doc is trying to explain things, I'm aware that out in the wild developers probably create a virtual service for each pod if they need granular traffic control to them instead of doing it on a namespace level.

rcaballeromx replied:
@oaktowner You can see the flow and how the hosts are intended to be resolved. IMU (In My Understanding), he wildcard in `hosts:` allow configured virtual service to apply to any pods (svc-X) in the `my-namespace.svc.cluster.local` namespace. Then the specific routes are configured for individual pods. The prefix is used because what distinguishes one pod's name from the others is what comes before.

rcaballeromx commented on content/docs/concepts/network/index.md:
```diff
@@ -0,0 +1,437 @@
+## Gateways
+
+The Istio gateways are layer 7 load balancers. With an Istio gateway, you can
+configure routes for HTTP and TCP traffic regardless of where the gateway runs.
```
Your confusion is warranted, As @frankbu explained, gateway here is referring to the actual workload, which can run in a number of places. In a k8s mesh it would be in a pod. I'm changing the term "gateway" to "gateway workload" to avoid this confusion.

rcaballeromx commented on content/docs/concepts/network/index.md:
```diff
@@ -0,0 +1,437 @@
+The Istio gateways are layer 7 load balancers. With an Istio gateway, you can
+configure routes for HTTP and TCP traffic regardless of where the gateway runs.
+Your mesh can have any number of gateways and multiple different gateway
+implementations can co-exist within your mesh. Using a gateway, you can
```
Done

rcaballeromx commented on content/docs/concepts/network/index.md:
```diff
@@ -0,0 +1,437 @@
+By separating the layer 4 and 6 specifications from the layer 7, the Istio
+gateways overcome the shortcomings of the layer 4 and 6 specifications. Istio
+only configures the functions that layer 7 load balancing proxies implement
+uniformly. To use standard Istio rules to control HTTP requests as well as TCP
```
Done.

rcaballeromx commented:
I've addressed all comments on the text. @frankbu @oaktowner Please let me know if any of the changes don't address your concerns with the text.
@rshriram The key difference between this content and the "Traffic management" concept is: 
Traffic management is about how to manage traffic for specific purposes.
Network configuration objects is about the configurable components available and how they work with one another.
They are certainly related and I've added cross-references between them to cement that relationship.
However, the documents address fundamentally different questions.

rcaballeromx commented:
@rshriram @frankbu @oaktowner Please review the text while I try to improve the diagrams based on @oaktowner 's comments.

geeknoid commented on content/docs/concepts/network/index.md:
```diff
@@ -0,0 +1,445 @@
+---
+title: Network configuration objects
```
We capitalize our titles, so "Network Configuration Objects"

geeknoid commented on content/docs/concepts/network/index.md:
```diff
@@ -0,0 +1,445 @@
+aliases:
+---
+
+## The Istio network configuration objects and your application
```
Please don't start with a title. The page as a whole already has a title rendered by the site infra.

geeknoid commented on content/docs/concepts/network/index.md:
```diff
@@ -0,0 +1,445 @@
+Lastly, the document discusses the benefits of using the Istio network API for
+your cloud applications.
+
+## The Istio network configuration objects
```
Somewhat redundant, given the title of the page.

geeknoid commented:
I'm definitely concerned about this doc, relative to the Traffic Management doc. The TM doc already discusses the same kinds of objects.

Right now, in our Concepts, we have one page per major functional area in the product. The idea is to give our customers a single unambiguous place to go to find the overall description of the area. This new doc now introduces two conceptual overview of the networking functionality, which I think is confusing and somewhat redundant.

When I initiatlly read drafts of this doc, I was expecting it to be a full replacement for the existing TM concept doc, and I still think that's the right path forward. I propose to merge this doc's new content into the TM doc, thus maintaining the idea of having a single conceptual overview of each functional area.

Thanks.

istio-testing commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/istio/istio.io/pull/2782#" title="Author self-approved">rcaballeromx</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[content/OWNERS](https://github.com/istio/istio.io/blob/master/content/OWNERS)~~ [rcaballeromx]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

rcaballeromx commented:
@frankbu @rshriram @geeknoid I have completed the integration of the new content with our existing Traffic management content. During that process, I had to restructure the entire section. You can see exactly how on the [commit message](https://github.com/istio/istio.io/pull/2782/commits/27c44739b5305566a3b71364a3a17b13b6e67bdc)
The new structure is clearer, easier to consume, and does not assume the content is read from beginning to end - all good things. 
PTAL

rcaballeromx commented:
/retest

frankbu commented on content/docs/concepts/traffic-management/routing-rules/index.md:
```diff
@@ -0,0 +1,271 @@
+
+{{< warning_icon >}} Consider the priorities of each rule in a configuration file when there is more than one.
+
+## Multiple match conditions {#multi-condition}
```
Multiple match conditions are "conditional rules", so this section belongs after the following one (possibly even as a subsection of it).

rcaballeromx replied:
Agreed, I'll move the content then.

rcaballeromx commented on content/docs/concepts/traffic-management/routing-rules/index.md:
```diff
@@ -0,0 +1,271 @@
+regardless of the configuration object where you configure them. If you use the
+examples provided, you must adapt them to your application and apply the
+configuration in the appropriate configuration object. For simplicity, all
+examples use [virtual services](../virtual-services).
```
Thank you for the clarification. To enhance clarity and keep the content modular, I'll keep this content as a separate file but will emphasize what you just clarified. I will link both the virtual services section to this one and vice-versa. One question remains: Are routing rules only found on virtual services? It is my understanding that destination rules can also contain them, is that correct?

frankbu replied:
No. Destination rules define the labels that identify service subsets and post-routing destination configuration. During routing, after a particular route rule (in a virtual service) is selected and consequently the exact destination is determined, then a destination rule configuration for that host (if one exists) is used to control policies, like circuit breakers, loadbalancing policy for the instances/pods implementing that service (or subset), and tls properties to use when calling that particular route destination.

frankbu commented on content/docs/concepts/traffic-management/destination-rules/index.md:
```diff
@@ -0,0 +1,180 @@
+Visit our [destination rules reference documentation](/docs/reference/config/istio.networking.v1alpha3/#DestinationRule)
+to review all the enabled keys and values.
+
+## Destination rules routing evaluation
```
Destination rules have nothing to do with routing. Maybe a better title could be "Destination rules evaluation", but see my following comment about the appropriateness of describing the evaluation algorithm in the Destination rules section of the doc.

frankbu commented on content/docs/concepts/traffic-management/destination-rules/index.md:
```diff
@@ -0,0 +1,180 @@
+to review all the enabled keys and values.
+
+## Destination rules routing evaluation
+
```
There is a pervasive incorrect characterization of the role of Pilot in many of these docs (i.e., not just in this section), but here specifically everything it says Pilot does is not correct. Pilot does NOT do any of these things. Pilot's job is to accept user configuration objects, validate them for correctness, and then distribute them to the appropriate Envoy proxies. Envoy does all the runtime evaluation based on the user's configuration it's been given.

frankbu commented on content/docs/concepts/traffic-management/destination-rules/index.md:
```diff
@@ -0,0 +1,180 @@
+## Destination rules routing evaluation
+
+Similar to route rules in virtual services, **Pilot** associates policies
+defined in a destination rule with a particular `host:` service. If the
```
Pilot doesn't associate policies with hosts, users do.

frankbu commented on content/docs/concepts/traffic-management/destination-rules/index.md:
```diff
@@ -0,0 +1,180 @@
+Similar to route rules in virtual services, **Pilot** associates policies
+defined in a destination rule with a particular `host:` service. If the
+policies are subset specific, the activation depends on the route rule
+evaluation results.
```
Before saying this, you would need to say that policies can optionally be subset-specific.

frankbu commented on content/docs/concepts/traffic-management/destination-rules/index.md:
```diff
@@ -0,0 +1,180 @@
+policies are subset specific, the activation depends on the route rule
+evaluation results.
+
+The rule evaluation process is as follows:
```
All 4 of the following steps are wrong. I also don't think that describing the rule evaluation algorithm in the Destination rule section of the docs makes sense.

frankbu commented on content/docs/concepts/traffic-management/destination-rules/index.md:
```diff
@@ -0,0 +1,180 @@
+The rule evaluation process is as follows:
+
+1. **Pilot** evaluates if the route rules in the virtual service configuration
+   files correspond to the requested `host:`.
```
Step 1. Envoy evaluates the routing rules of a configured virtual service corresponding to the incoming request host, if there is one.
Step 2. If no rule matching the request is found, the request falls through to lower-level round-robin routing.
Step 3. Otherwise, the matching rule's destination host (and optionally subset) is used to locate a destination rule policies that should be applied to the request, if there is one defined.
Step 4. Envoy then routes the request to the selected destination using the destination rule policies that it found, or default policies otherwise.

LisaFC commented on content/docs/concepts/traffic-management/pilot/index.md:
```diff
@@ -0,0 +1,96 @@
+best fulfills your needs. You can configure routing rules using any of the four
+network configuration objects in Istio:
+
+* A [virtual service](../virtual-services) configures the routing rules
```
I'd be a bit stronger here, virtual services are kind of core to how the whole traffic management model works. Maybe something like

, which uses routing rules to describe the mapping between one or more user-addressable destinations to the actual destination workloads inside the mesh

LisaFC commented on content/docs/concepts/traffic-management/traffic-routing/index.md:
```diff
@@ -0,0 +1,135 @@
+---
+title: Traffic Routing
```
I'm not sure about this title. Most users will think of what you do with Virtual Services as traffic routing (ie specifying exactly which destination services you want to get which traffic), which is not what this doc is about. Service subsets? Grouping services?

LisaFC commented on content/docs/concepts/traffic-management/traffic-routing/index.md:
```diff
@@ -0,0 +1,135 @@
+various fields from the metadata in the platform.
+
+To cover common configuration scenarios, Istio introduces the concept of a
+**service subset**, to subdivide service instances by versions, say `v1` and
```
**service subset**. Defining subsets of services lets you subdivide your service instances by versions....

LisaFC commented on content/docs/concepts/traffic-management/traffic-routing/index.md:
```diff
@@ -0,0 +1,135 @@
+`v2`, or environment, say `staging` and `prod`. These variants are not
+necessarily different API versions: they can be iterative changes to the same
+service, deployed in different environments like prod, staging, dev, etc.
+Together with Istio's network configuration objects, you can use subsets to
```
Once you've defined your service subsets, you can use these subsets in Istio's network configuration objects, for example to direct a particular type of traffic to a particular service version in a [virtual service](), or to specify security policies for a particular service version in a [destination rule]().

LisaFC commented on content/docs/concepts/traffic-management/traffic-routing/index.md:
```diff
@@ -0,0 +1,135 @@
+provide additional control over the traffic between services.
+
+This concept covers the common configuration scenarios for services subsets,
+like:
```
This page describes some of the scenarios where you might want to use service subsets, including:

LisaFC commented on content/docs/concepts/traffic-management/traffic-routing/index.md:
```diff
@@ -0,0 +1,135 @@
+- [Splitting traffic between versions](#splitting) for A/B testing
+- [Circuit breakers](../failures/#circuit)
+
+## Communication between services {#communication}
```
Again, odd title - having service subsets doesn't reeeeeally allow communication between services, but the Istio config/traffic model does allow for easy communication between services that have multiple versions/subsets.  Maybe "Subsets and Virtual Services"? This whole section is a bit of a repeat of some of the stuff we've covered in the other pages, particularly around virtual services. I'd be tempted to take it out, move some of it elsewhere, or cut it down dramatically and link to the virtual services and Pilot/Envoy pages for details.

LisaFC commented on content/docs/concepts/traffic-management/traffic-routing/index.md:
```diff
@@ -0,0 +1,135 @@
+
+## Splitting traffic between versions {#splitting}
+
+Each route rule identifies one or more weighted workloads to call when the rule
```
Each route rule in a [virtual service]...

LisaFC commented on content/docs/concepts/traffic-management/traffic-routing/index.md:
```diff
@@ -0,0 +1,135 @@
+Each route rule identifies one or more weighted workloads to call when the rule
+is activated. Each workload corresponds to a specific version of the
+destination service. You use subsets to express the versions. If there are
+multiple registered instances with the specified subset, **Pilot** routes them
```
Pilot -> Envoy

LisaFC commented on content/docs/concepts/traffic-management/traffic-routing/index.md:
```diff
@@ -0,0 +1,135 @@
+      weight: 25
+{{< /text >}}
+
+To learn more about configuring destination rules, visit our [Destination Rules concept](../destination-rules).
```
Shouldn't we be sending them to the virtual services page, not the destination rules page?

LisaFC commented on content/docs/concepts/traffic-management/virtual-services/index.md:
```diff
@@ -0,0 +1,168 @@
+keywords: [routing, request, traffic routing, requests routing, virtual service, subset, routing rules]
+aliases:
+---
+
```
As per comment above, I think we should combine this and the routing rules doc.

LisaFC replied:
Also I have a few more edits to do on this doc but I need to catch a train and I'm aware this is getting a bit epic! :)

LisaFC commented on content/docs/concepts/traffic-management/virtual-services/index.md:
```diff
@@ -0,0 +1,168 @@
+
+You can use virtual services for fine-grained traffic control. Istio can use
+virtual services to configure ingress, canary, traffic rewrite, or traffic
+policing. A virtual service is not required for simple connectivity. A service
```
You can use virtual services for fine-grained traffic control, including configuring ingress traffic and routing to specific service versions. However, you don't require a virtual service for simple connectivity.

LisaFC commented on content/docs/concepts/traffic-management/virtual-services/index.md:
```diff
@@ -0,0 +1,168 @@
+virtual services to configure ingress, canary, traffic rewrite, or traffic
+policing. A virtual service is not required for simple connectivity. A service
+can map one or more user-addressable destinations to the actual destination
+workloads inside the mesh without a virtual service. The Istio virtual services
```
"A service can...."

Is this right?

LisaFC replied:
Using virtual services you can configure traffic routes to:

LisaFC commented on content/docs/concepts/traffic-management/virtual-services/index.md:
```diff
@@ -0,0 +1,168 @@
+
+As you can see, virtual services are very flexible and can accommodate a wide
+variety of topologies. Virtual services configure the [traffic routes](../traffic-routing)
+to your application services. The Istio virtual services can use any DNS names
```
Virtual services configure...

remove this sentence, redundant.

rcaballeromx commented:
@frankbu Thank you for taking a look again. I'll work with @LisaFC to address what has become incorrect or inaccurate through my changes. However, let me explain why and how the new content introduced inaccuracies. The original content is accurate by providing little detail and remaining vague. My being more explicit and clear in the new content causes those flaws to surface.
I can correct the inaccuracies you find in the new content and I can change the structure to fix any flow issues but the original content requires a lot more than just minor tweaks.

istio-testing commented:
@rcaballeromx: PR needs rebase.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

geeknoid commented:
@rcaballeromx Do you still intend to move forward with this PR?

geeknoid commented:
I'm closing this PR. Please reopen if you decide to move forward with it.

rcaballeromx commented:
That is fine, the next iteration does not need this PR. Thanks for closing @geeknoid
